pkg/ruler/merger.go

```diff
@@ -9,26 +10,51 @@ import (
 // mergeGroupStateDesc removes duplicates from the provided []*GroupStateDesc by keeping the GroupStateDesc with the
 // latest information. It uses the EvaluationTimestamp of the GroupStateDesc and the EvaluationTimestamp of the
 // ActiveRules in a GroupStateDesc to determine the which GroupStateDesc has the latest information.
-func mergeGroupStateDesc(in []*GroupStateDesc) []*GroupStateDesc {
+// It also truncates rule groups if maxRuleGroups > 0
+func mergeGroupStateDesc(ruleResponses []*RulesResponse, maxRuleGroups int32, dedup bool) *RulesResponse {
+
+	var groupsStateDescs []*GroupStateDesc
+
+	for _, resp := range ruleResponses {
+		groupsStateDescs = append(groupsStateDescs, resp.Groups...)
+	}
+
 	states := make(map[string]*GroupStateDesc)
 	rgTime := make(map[string]time.Time)
-	for _, state := range in {
-		latestTs := state.EvaluationTimestamp
-		for _, r := range state.ActiveRules {
-			if latestTs.Before(r.EvaluationTimestamp) {
-				latestTs = r.EvaluationTimestamp
+	groups := make([]*GroupStateDesc, 0)
+	if dedup {
+		for _, state := range groupsStateDescs {
+			latestTs := state.EvaluationTimestamp
+			for _, r := range state.ActiveRules {
+				if latestTs.Before(r.EvaluationTimestamp) {
+					latestTs = r.EvaluationTimestamp
+				}
 			}
+			key := promRules.GroupKey(state.Group.Namespace, state.Group.Name)
+			ts, ok := rgTime[key]
+			if !ok || ts.Before(latestTs) {
+				states[key] = state
+				rgTime[key] = latestTs
+			}
+		}
+		for _, state := range states {
+			groups = append(groups, state)
 		}
-		key := promRules.GroupKey(state.Group.Namespace, state.Group.Name)
-		ts, ok := rgTime[key]
-		if !ok || ts.Before(latestTs) {
-			states[key] = state
-			rgTime[key] = latestTs
+	} else {
+		groups = groupsStateDescs
+	}
+
+	if maxRuleGroups > 0 {
+		//Need to sort here before we truncate
+		sort.Sort(GroupStateDescs(groups))
```

Comment:
I wonder if we should always sort the result.

We are already sorting the results in the prometheus list rules: https://github.com/cortexproject/cortex/blob/c25b18d514a191182a818c8f0c954564cf6ceaf4/pkg/ruler/api.go#L243

So we could sort just once here instead for all the cases. This will reduce modality and simplify the code a bit. WDYT @yeya24 ?

Reply:
Also, where is the next token from the previous request being used here to generate the response starting from the rule group that we truncated?

Reply:
> I wonder if we should always sort the result.
> 
> We are already sorting the results in the prometheus list rules:
> 
> https://github.com/cortexproject/cortex/blob/c25b18d514a191182a818c8f0c954564cf6ceaf4/pkg/ruler/api.go#L243
> 
> So we could sort just once here instead for all the cases. This will reduce modality and simplify the code a bit. WDYT @yeya24 ?

Without pagination, the size of the list can be pretty substantial and we would sort it twice. This was the reason I did not want to sort for non-paginated requests

Reply:
> Also, where is the next token from the previous request being used here to generate the response starting from the rule group that we truncated?

Each response from getLocalRules has been filtered using the next token

Reply:
> Each response from getLocalRules has been filtered using the next token

but what about the case you are using sharding?

Reply:
Ok, we discussed offline, and there was a misunderstanding from my part. I stand corrected.

Since we are truncating on a sorted list, where the sorting happens by the sha(namespace + group), it is safe to use the `>=` as filter when we spread the requests to get the rules from other rulers. This is like a distributed filter operation.

Having said that, It would not hurt to add a bit of description to the PR to detail how the implementation was done.